examples/seismic/skew_self_adjoint/wavesolver.py

```diff
@@ -0,0 +1,312 @@
+                                     save=save, **self._kwargs)
+
+        rec.data[:] = 0
+        u0.data[:] = 0
+        du.data[:] = 0
+        summary = op.apply(dm=dm, u0=u0, du=du, **kwargs)
+        return rec, u0, du, summary
+
+    def jacobian_adjoint(self, rec, u0, b=None, v=None, wOverQ=None,
+                         dm=None, du=None, save=None, **kwargs):
+        """
+        Linearized JacobianForward modeling function that creates the necessary
+        data objects for running a Jacobian forward modeling operator.
+        Required parameters: rec, u0.
+
+        Parameters
+        ----------
+        rec : SparseTimeFunction, required
+            The interpolated receiver data to be injected.
+        u0 : Function or float, required, (created with save=True)
+            Stores the computed background wavefield.
+        b : Function or float, optional, defaults to b at construction
+            The time-constant buoyancy.
+        v : Function or float, optional, defaults to v at construction
+            The time-constant velocity.
+        wOverQ : Function or float, optional, defaults to wOverQ at construction
+            The time-constant dissipation only attenuation w/Q field.
+        dm : Function or float, optional, defaults to new Function
+            The perturbation to the velocity model.
+        du : Function or float, optional, defaults to new TimeFunction
+            Stores the computed perturbed wavefield.
+        save : int or Buffer, optional
+            The entire (unrolled) wavefield.
+
+        Returns
+        ----------
+        Function model perturbation dm, Receiver time series data rec,
+        TimeFunction background wavefield u0, TimeFunction perturbation wavefield du,
+        and performance summary
+        """
+        # Get model perturbation Function or create
+        dm = dm or Function(name='dm', grid=self.v.grid, space_order=self.space_order)
+
+        # Get (b, v, wOverQ) from passed arguments or from (b, v, wOverQ) at construction
+        b = b or self.b
+        v = v or self.v
+        wOverQ = wOverQ or self.wOverQ
+
+        # ensure rec, b, v, wOverQ all share the same underlying grid
+        assert rec.grid == b.grid == v.grid == wOverQ.grid
+
+        # Make dictionary of the physical model properties
+        model = {'b': b, 'v': v, 'wOverQ': wOverQ}
+
+        # Create the perturbation wavefield if not provided
+        du = du or TimeFunction(name='du', grid=self.v.grid,
+                                save=self.time_axis.num if save else None,
+                                time_order=2, space_order=self.space_order)
+
+        # Execute operator, "splatting" the model dictionary entries
+        op = ISO_JacobianAdjOperator(model, rec, self.time_axis,
+                                     space_order=self.space_order,
+                                     save=save, **self._kwargs)
+
+        dm.data[:] = 0
+        du.data[:] = 0
```

Comment:
? this defeat the purpose of having it as an input, in particualr for dm where for example you could pass it the result from previous source as input to kepp stacking into it.

Reply:
left from debugging, gone now